pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,167 @@
+	ss.wObjSync[id] = true
+}
+
+func (ss *singletonState) Unset(id util.ObjectIdentifier) {
+	ss.wObjSync[id] = false
+}
+
+// Should it be a field of Contoller?
+var theSingletonState singletonState = singletonState{
+	wObjSync: map[util.ObjectIdentifier]bool{},
+}
+
+func (c *Controller) buildSingletonStateAndOptionallyReconcile(ctx context.Context, reconcile bool) error {
+	logger := klog.FromContext(ctx)
+	logger.Info("Initializing the desired state for singleton statuses")
+
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, bdg := range allBdgs.Items {
+		logger.V(2).Info("Got Binding", "name", bdg.Name)
+		list := bdg.Spec.Workload.NamespaceScope // Should also do the same thing for cluster scoped
+		for _, item := range list {
+			nsWObjRef := item.NamespaceScopeDownsyncObject
+			lister, found := c.listers.Get(metav1GVR2schemaGVR(nsWObjRef.GroupVersionResource))
+			if !found {
+				return fmt.Errorf("could not get lister for gvr: %s", nsWObjRef.GroupVersionResource)
+			}
+			nsWObj, err := lister.ByNamespace(nsWObjRef.Namespace).Get(nsWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := nsWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				logger.V(2).Info("Found namespace scoped singleton workload object", "name", nsWObjRef.Name)
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: nsWObjRef.Group, Version: nsWObjRef.Version, Kind: nsWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   nsWObjRef.Resource,
+					ObjectName: cache.NewObjectName(nsWObjRef.Namespace, nsWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should has status synced", "name", nsWObjRef.Name)
+					theSingletonState.Set(wObjIdentifier)
+				} else {
+					logger.V(2).Info("Singleton workload object should not have status synced", "name", nsWObjRef.Name)
+					theSingletonState.Unset(wObjIdentifier)
+				}
+				// reconcile should be out of the for loop over all bindings
+				// because the desired state for a single workload object must be determined over ALL bindings
+				if reconcile {
+					c.reconcileSingletonByWObj(ctx, wObjIdentifier)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func metav1GVR2schemaGVR(in metav1.GroupVersionResource) (out schema.GroupVersionResource) {
+	out = schema.GroupVersionResource{
+		Group:    in.Group,
+		Version:  in.Version,
+		Resource: in.Resource,
+	}
+	return
+}
+
+func (c *Controller) reconcileSingletonByWObj(ctx context.Context, wObjID util.ObjectIdentifier) error {
+	logger := klog.FromContext(ctx)
+
+	if !theSingletonState.Get(wObjID) {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &wObjID, emptyStatus, c.listers, c.wdsDynClient)
```

Comment:
I think this is too pushy, it will overwrite contents of the Status section that were written by other controllers. Instead this should consult the `ObjectMeta.ManagedFields` to find the `Status` contents that were written by this controller and delete only those.

Reply:
I agree. Can we address the status ownership issue in another PR? I documented this as https://github.com/kubestellar/kubestellar/issues/2393